### ISSUE TITLE
MP JWT 1.2 TCK RequiredClaimsTest for verifyTokenWithoutName fails with 404 when expecting 401

### ISSUE THREAD

**Issue by clarkek123** [OpenLiberty/open-liberty#13757]
**Describe the bug**
The MP JWT 1.2 TCK RequiredClaimsTest for verifyTokenWithoutName is failing with a 404 when it is expecting a 401.

If there is a stack trace, please include the FULL stack trace (without any `[internal classes]` lines in it). To find the full stack trace, you may need to check in `$WLP_OUTPUT_DIR/messages.log`

**Steps to Reproduce**
1. Run MP JWT 1.2 TCK tests

**Expected behavior**
TCK test expects 401 return code and is getting a 404 along with an unexpected message and FFDC logged.

**Diagnostic information:**
 - OpenLiberty Version: development master branch 
 - Java Version: 

Eclipse OpenJ9 VM AdoptOpenJDK (build openj9-0.17.0, JRE 11 Mac OS X amd64-64-Bit Compressed References 20191016_371 (JIT enabled, AOT enabled)
OpenJ9   - 77c1cf708
OMR      - 20db4fbc
JCL      - 2a7af5674b based on jdk-11.0.5+10)

**Additional context**
Test:  RequiredClaimsTest
New test added with mp-jwt 1.2 TCK: verifyTokenWithoutName

This test fails because it expects to get a 401 but instead is getting a RC 404. It also gets an unexpected message and FFDC.

The test source shows the following is being tested for the verifyTokenWithoutName test:

```Java
    @RunAsClient
    @Test(groups = TEST_GROUP_JWT,
            description = "Verify that HTTP 401 status is returned if the token contains no 'upn', 'preferred_username' and 'sub' claims")
    public void verifyTokenWithoutName() throws Exception {
        String tokenWithoutName = TokenUtils.signClaims("/TokenWithoutName.json");
        Reporter.log("Begin verifyTokenWithoutName\n");
        String uri = baseURL.toExternalForm() + "endp/verifyTokenWithoutName";
        WebTarget echoEndpointTarget = ClientBuilder.newClient()
                .target(uri);
        Response response = echoEndpointTarget.request(MediaType.APPLICATION_JSON).header(HttpHeaders.AUTHORIZATION,
            "Bearer " + tokenWithoutName).get();
        Assert.assertEquals(response.getStatus(), HttpURLConnection.HTTP_UNAUTHORIZED);
    }
```

The messages.log shows:
```
[8/31/20, 16:24:58:883 CDT] 00000033 com.ibm.ws.webcontainer.servlet                              I SRVE0242I: [RequiredClaimsTest] [/RequiredClaimsTest] [org.eclipse.microprofile.jwt.tck.container.jaxrs.TCKApplication]: Initialization successful.
[8/31/20, 16:24:58:914 CDT] 00000033 y.authorization.builtin.internal.BuiltinAuthorizationService I CWWKS2104I: The authorization decision for resources in application RequiredClaimsTest will be made by using the group names of the user that matches the role names required to access that resource.
[8/31/20, 16:25:00:649 CDT] 0000003a com.ibm.ws.security.mp.jwt.tai.MicroProfileJwtTAI            E CWWKS5523E: The MicroProfile JWT feature cannot authenticate the request because the token that is included in the request cannot be validated. CWWKS6044E: The JSON Web Token (JWT) is not valid because the issued at ('iat') claim specifies a date later than the current time. The 'iat' claim time is [2020-08-31T16:30:05-0500]. The current time plus the clock skew is [2020-08-31T16:30:00-0500]. The configured clock skew is [300] seconds.
[8/31/20, 16:25:00:774 CDT] 0000003a com.ibm.ws.security.mp.jwt.tai.MicroProfileJwtTAI            E CWWKS5523E: The MicroProfile JWT feature cannot authenticate the request because the token that is included in the request cannot be validated. CWWKS6044E: The JSON Web Token (JWT) is not valid because the issued at ('iat') claim specifies a date later than the current time. The 'iat' claim time is [2020-08-31T16:30:05-0500]. The current time plus the clock skew is [2020-08-31T16:30:00-0500]. The configured clock skew is [300] seconds.
[8/31/20, 16:25:00:856 CDT] 00000035 com.ibm.ws.security.mp.jwt.tai.MicroProfileJwtTAI            E CWWKS5523E: The MicroProfile JWT feature cannot authenticate the request because the token that is included in the request cannot be validated. CWWKS6025E: The JSON Web Token (JWT) is not valid because its expiration ('exp') claim is either missing or the token expired. The expiration claim is [null]. The current time minus the clock skew is [2020-08-31T16:20:00-0500]. The configured clock skew is [300] seconds.
[8/31/20, 16:25:00:908 CDT] 00000035 com.ibm.ws.security.mp.jwt.tai.MicroProfileJwtTAI            E CWWKS5523E: The MicroProfile JWT feature cannot authenticate the request because the token that is included in the request cannot be validated. CWWKS6025E: The JSON Web Token (JWT) is not valid because its expiration ('exp') claim is either missing or the token expired. The expiration claim is [null]. The current time minus the clock skew is [2020-08-31T16:20:00-0500]. The configured clock skew is [300] seconds.
[8/31/20, 16:25:01:000 CDT] 0000003f com.ibm.ws.security.mp.jwt.impl.utils.JwtPrincipalMapping    E CWWKS5519E: The authenticated user information does not contain the claim [upn] that is specified by the [userNameAttribute] attribute in the MicroProfile JWT configuration.
[8/31/20, 16:25:01:000 CDT] 0000003f com.ibm.ws.security.mp.jwt.tai.TAIMappingHelper              E CWWKS5506E: The user name cannot be extracted from the token.
[8/31/20, 16:25:01:028 CDT] 0000003f com.ibm.ws.logging.internal.impl.IncidentImpl                I FFDC1015I: An FFDC Incident has been created: "com.ibm.ws.security.mp.jwt.error.MpJwtProcessingException: CWWKS5506E: The user name cannot be extracted from the token. com.ibm.ws.security.mp.jwt.tai.MicroProfileJwtTAI 387" at ffdc_20.08.31_16.25.01.0.log
[8/31/20, 16:25:01:031 CDT] 0000003f com.ibm.ws.security.mp.jwt.tai.MicroProfileJwtTAI            E CWWKS5508E: The MicroProfile JWT feature cannot create a subject for the user with the provided token by using MicroProfile JWT configuration [defaultMpJwt]. CWWKS5506E: The user name cannot be extracted from the token.
[8/31/20, 16:25:01:928 CDT] 00000037 com.ibm.ws.config.xml.internal.ConfigRefresher               A CWWKG0016I: Starting server configuration update.
[8/31/20, 16:25:01:930 CDT] 00000037 com.ibm.ws.config.xml.internal.XMLConfigParser               A CWWKG0028A: Processing included configuration resource: /Users/kristenclarke/libertyGit/open-liberty/dev/build.image/wlp/usr/servers/fatTestPorts.xml
[8/31/20, 16:25:01:931 CDT] 00000037 com.ibm.ws.config.xml.internal.XMLConfigParser               A CWWKG0028A: Processing included configuration resource: /Users/kristenclarke/libertyGit/open-liberty/dev/build.image/wlp/usr/servers/fatTestCommon.xml
[8/31/20, 16:25:01:940 CDT] 00000037 com.ibm.ws.config.xml.internal.ConfigRefresher               A CWWKG0017I: The server configuration was successfully updated in 0.011 seconds.
[8/31/20, 16:25:01:989 CDT] 0000003f com.ibm.ws.http.internal.VirtualHostImpl                     A CWWKT0017I: Web application removed (default_host): http://192.168.1.19:8010/RequiredClaimsTest/
```

The unexpected message is as follows. This looks like it is probably OK and I should update to expect this message. Please confirm.
CWWKS5519E: The authenticated user information does not contain the claim [upn] that is specified by the [userNameAttribute] attribute in the MicroProfile JWT configuration.

The unexpected FFDC is as follows. Is it OK to get an FFDC for this case? Please confirm.

```
------Start of DE processing------ = [8/31/20, 16:25:01:026 CDT]
Exception = com.ibm.ws.security.mp.jwt.error.MpJwtProcessingException
Source = com.ibm.ws.security.mp.jwt.tai.MicroProfileJwtTAI
probeid = 387
Stack Dump = com.ibm.ws.security.mp.jwt.error.MpJwtProcessingException: CWWKS5506E: The user name cannot be extracted from the token.
	at com.ibm.ws.security.mp.jwt.tai.TAIMappingHelper.setUsername(TAIMappingHelper.java:156)
	at com.ibm.ws.security.mp.jwt.tai.TAIMappingHelper.<init>(TAIMappingHelper.java:72)
	at com.ibm.ws.security.mp.jwt.tai.MicroProfileJwtTAI.createResult(MicroProfileJwtTAI.java:432)
	at com.ibm.ws.security.mp.jwt.tai.MicroProfileJwtTAI.handleMicroProfileJwtValidation(MicroProfileJwtTAI.java:413)
	at com.ibm.ws.security.mp.jwt.tai.MicroProfileJwtTAI.getAndValidateMicroProfileJwt(MicroProfileJwtTAI.java:373)
	at com.ibm.ws.security.mp.jwt.tai.MicroProfileJwtTAI.handleRequestBasedOnJwtConfig(MicroProfileJwtTAI.java:315)
	at com.ibm.ws.security.mp.jwt.tai.MicroProfileJwtTAI.getAssociatedConfigAndHandleRequest(MicroProfileJwtTAI.java:293)
	at com.ibm.ws.security.mp.jwt.tai.MicroProfileJwtTAI.negotiateValidateandEstablishTrust(MicroProfileJwtTAI.java:266)
	at com.ibm.ws.webcontainer.security.internal.TAIAuthenticator.authenticate(TAIAuthenticator.java:126)
	at com.ibm.ws.webcontainer.security.WebProviderAuthenticatorProxy.handleTAI(WebProviderAuthenticatorProxy.java:333)
	at com.ibm.ws.webcontainer.security.WebProviderAuthenticatorProxy.authenticate(WebProviderAuthenticatorProxy.java:450)
	at com.ibm.ws.webcontainer.security.WebAuthenticatorProxy.authenticate(WebAuthenticatorProxy.java:68)
	at com.ibm.ws.webcontainer.security.WebAppSecurityCollaboratorImpl.authenticateRequest(WebAppSecurityCollaboratorImpl.java:1211)
	at com.ibm.ws.webcontainer.security.WebAppSecurityCollaboratorImpl.setAuthenticatedSubjectIfNeeded(WebAppSecurityCollaboratorImpl.java:833)
	at com.ibm.ws.webcontainer.security.WebAppSecurityCollaboratorImpl.optionallyAuthenticateUnprotectedResource(WebAppSecurityCollaboratorImpl.java:790)
	at com.ibm.ws.webcontainer.security.WebAppSecurityCollaboratorImpl.determineWebReply(WebAppSecurityCollaboratorImpl.java:960)
	at com.ibm.ws.webcontainer.security.WebAppSecurityCollaboratorImpl.performSecurityChecks(WebAppSecurityCollaboratorImpl.java:656)
	at com.ibm.ws.webcontainer.security.WebAppSecurityCollaboratorImpl.preInvoke(WebAppSecurityCollaboratorImpl.java:574)
	at com.ibm.wsspi.webcontainer.collaborator.CollaboratorHelper.preInvokeCollaborators(CollaboratorHelper.java:459)
	at com.ibm.ws.webcontainer.osgi.collaborator.CollaboratorHelperImpl.preInvokeCollaborators(CollaboratorHelperImpl.java:270)
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1114)
	at com.ibm.ws.webcontainer.webapp.WebApp.handleRequest(WebApp.java:5057)
	at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.handleRequest(DynamicVirtualHost.java:314)
	at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:1005)
	at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:279)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:1134)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.wrapHandlerAndExecute(HttpDispatcherLink.java:415)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.ready(HttpDispatcherLink.java:374)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleDiscrimination(HttpInboundLink.java:577)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleNewRequest(HttpInboundLink.java:511)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.processRequest(HttpInboundLink.java:359)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.ready(HttpInboundLink.java:326)
	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.sendToDiscriminators(NewConnectionInitialReadCallback.java:167)
	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.complete(NewConnectionInitialReadCallback.java:75)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.requestComplete(WorkQueueManager.java:504)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.attemptIO(WorkQueueManager.java:574)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.workerRun(WorkQueueManager.java:958)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager$Worker.run(WorkQueueManager.java:1047)
	at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:239)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:831)
```

**Comment by ayoho** [comment]
This one actually looks like a bug in the TCK to me. All of the tests invoke a JAX-RS endpoint specifically written for each test. Each of those endpoints is configured to do checking that’s appropriate for the respective test. See:

```Java
    @RunAsClient
    @Test(groups = TEST_GROUP_JWT,
            description = "Verify that the exp claim is as expected")
    public void verifyExpiration() throws Exception {
        ...
        String uri = baseURL.toExternalForm() + "endp/verifyExpiration";
        ...
    }
    ...
    @RunAsClient
    @Test(groups = TEST_GROUP_JWT,
            description = "Verify that HTTP 401 status is returned if the token contains no 'upn', 'preferred_username' and 'sub' claims")
    public void verifyTokenWithoutName() throws Exception {
        ...
        String uri = baseURL.toExternalForm() + "endp/verifyTokenWithoutName";
        ...
    }
```

So
- `verifyExpiration` invokes `endp/verifyExpiration`
- `verifyTokenWithoutExpiration` invokes `endp/verifyTokenWithoutExpiration`
- `verifyTokenWithoutName` invokes `endp/verifyTokenWithoutName`
- So on for all the other tests

The problem is that if you look at https://github.com/eclipse/microprofile-jwt-auth/blob/f2a0557d1ec0544b9c404216ff847dc401814fb4/tck/src/test/java/org/eclipse/microprofile/jwt/tck/container/jaxrs/RequiredClaimsEndpoint.java#L43 - the endpoint class that defines all of those endpoints - it doesn’t have a “/verifyTokenWithoutName” endpoint defined.

e.g.:
```Java
    @GET
    @Path("/verifyExpiration")
    @Produces(MediaType.APPLICATION_JSON)
    public JsonObject verifyExpiration(@QueryParam("exp") Long exp) {
    ...
    @GET
    @Path("/verifyTokenWithoutExpiration")
    public Response verifyTokenWithoutExpiration() {
    ...
```

No sign of “verifyTokenWithoutName” anywhere in that class.

Our MP JWT code is doing everything correctly and returning back a 401. However the test will ultimately end up with a 404 when the request is served because the URL that's being accessed doesn't actually exist in the test application.

**Comment by ayoho** [comment]
This should be tracked by https://github.com/eclipse/microprofile-jwt-auth/issues/209

**Comment by ayoho** [comment]
https://github.com/eclipse/microprofile-jwt-auth/issues/209 has been resolved, so once we can pick up a new snapshot from the TCK, this test should work again.

**Comment by clarkek123** [comment]
This is fixed with the MP JWT TCK 1.2 - RC2 which I've just enabled in automation. I'm closing this issue.
